### PR TITLE
added filesize overflow check for mmap on 32bit os

### PR DESCRIPTION
**Pull request by c3V6a2Vy**
This has been causing problems with my 32-bit ARMv7L machine. 

It seems like aria2 is compiled with 64-bit offset support so the `filesize` (`st_size` of [`stat`](http://linux.die.net/man/2/stat)) would be a 64-bit `off_t`, even in a 32-bit O/S. 

The problem is, `length` arg in [`mmap`](http://linux.die.net/man/2/mmap) is a 32-bit `size_t` in 32-bit O/S, so the input, `filesize`, would be truncated and the mmap would not be mapping into a correct size of address space.

This will cause `SIGSEGV` in later `memcpy` operations.

### CODE REVIEW

File: src/AbstractDiskWriter.cc [hunk 1]
```diff
@@ -373,6 +373,14 @@ void AbstractDiskWriter::ensureMmapWrite(size_t len, int64_t offset)
         return;
       }
 
+      uint32_t filesize_lo = filesize & 0xffffffffu;
+      if (sizeof(off_t) > sizeof(size_t) && (filesize_lo) > 0) {
```
Comment (tatsuhiro-t) [inline src/AbstractDiskWriter.cc]: I may be missing something, but we want to check filesize is larger than size_t right?  Then is `static_cast<uint64_t>(std::numeric_limits<size_t>::max()) < static_cast<uint64_t>(filesize)` enough?

**Comment by tatsuhiro-t** [comment]
Thank you for the PR.  Could you check my comment?

File: src/AbstractDiskWriter.cc (hunk 1 above)
Comment (c3V6a2Vy) [inline src/AbstractDiskWriter.cc]: You are right, I want to check if filesize is greater than max of size_t.

Your change looks more standardized. I will make the change.

**Comment by c3V6a2Vy** [comment]
found one more problem in the later `mmap` check. 

Instead of returning `null`, `mmap` will return the `MAP_FAILED` flag, which is `(void * )-1`.

> Return Value
> 
> On success, mmap() returns a pointer to the mapped area. On error, the value MAP_FAILED (that is, (void *) -1) is returned, and errno is set appropriately. On success, munmap() returns 0, on failure -1, and errno is set (probably to EINVAL).

**Comment by tatsuhiro-t** [comment]
Thank you!  Merged now.

**Pull request merged**